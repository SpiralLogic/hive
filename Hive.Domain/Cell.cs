using System.Collections.Generic;

namespace Hive.Domain
{
    public class Cell
    {
        public Cell(GameCoordinate position, string color)
        {
            Position = position;
            Color = color;
        }

        public GameCoordinate Position { get; }
        public string Color { get; }
        public ICollection<Tile> Tiles { get; } = new List<Tile>();
    }
}