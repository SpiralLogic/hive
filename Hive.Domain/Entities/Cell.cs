﻿using System.Collections.Generic;

namespace Hive.Domain.Entities
{
    public class Cell
    {
        public Cell(GameCoordinate coordinates, ICollection<Tile> tiles = null)
        {
            Coordinates = coordinates;
            Tiles = tiles ??  new List<Tile>();
        }

        public GameCoordinate Coordinates { get; }
        public ICollection<Tile> Tiles { get; }
    }
}