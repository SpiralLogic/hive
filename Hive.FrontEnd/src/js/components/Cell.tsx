import * as React from 'react';
import { useDrop } from 'react-dnd';
import { Hexagon, HexCoordinates } from '../domain';
import { Tile, TILE_TYPE } from './Tile';

type CellProps = Hexagon

const areEqual = (a: HexCoordinates, b: HexCoordinates) => a && b && a.q === b.q && a.r === b.r;
const isValidMove = (move: HexCoordinates, validMoves: HexCoordinates[]) => validMoves.some(dest => areEqual(move, dest));
const coordinateToString = ({ q, r }: HexCoordinates) => `${q}-${r}`;

export const Cell: React.FunctionComponent<CellProps> = ({
    tiles,
    coordinates,
}) => {
    const [{ isOver, canDrop }, drop] = useDrop({
        accept: TILE_TYPE,
        drop: () => coordinates,
        canDrop: (item, monitor) => isValidMove(coordinates, monitor.getItem().availableMoves),
        collect: monitor => ({
            isOver: monitor.isOver(),
            canDrop: monitor.canDrop(),
        }),
    });

    const backGroundColor = () => {
        if (canDrop && isOver) return ' valid-cell active';
        if (isOver) return ' invalid-cell';
        if (canDrop) return ' valid-cell';
        return '';
    };
    const id = coordinateToString(coordinates);
    return (
        <div className={'hex cell' + backGroundColor()} ref={drop}>
            {tiles.length > 0 && <Tile key={'tile' + id} {...tiles[0]} />}
        </div>
    );
};

Cell.displayName = 'Cell';