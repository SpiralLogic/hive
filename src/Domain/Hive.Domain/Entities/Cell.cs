using System;
using System.Collections.Generic;
using System.Linq;

namespace Hive.Domain.Entities
{
    public class Cell : IEquatable<Cell>
    {
        public Coords Coords { get; init; }
        public Stack<Tile> Tiles { get; init; } = new Stack<Tile>();
     
        public Cell(Coords coords)
        {
            Coords = coords;
        }

        public Cell AddTile(Tile tile)
        {
            Tiles.Push(tile);
            return this;
        }

        public bool IsEmpty() => !Tiles.Any();

        public Tile TopTile() => Tiles.Peek();

        public Tile RemoveTopTile() => Tiles.Pop();

        public bool Equals(Cell? other)
        {
            if (other is null) return false;
            if (ReferenceEquals(this, other)) return true;
            return Coords.Equals(other.Coords);
        }

        public override bool Equals(object? obj)
        {
            if (obj is null) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;
            return Equals((Cell)obj);
        }

        public override int GetHashCode()
        {
            return Coords.GetHashCode();
        }
    }
}