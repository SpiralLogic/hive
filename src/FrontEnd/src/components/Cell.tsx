import {Cell, HexCoordinates, Tile as TileType} from '../domain';
import {FunctionComponent, h} from 'preact';
import {TileDragEvent, useCellDropEmitter, useTileDragEmitter,} from '../emitters';
import {deepEqual} from 'fast-equals';
import {handleDragOver} from '../handlers';
import {memo} from 'preact/compat';
import {useEffect, useState} from 'preact/hooks';
import Tile from './Tile';

type Props = Cell;

const CellFC: FunctionComponent<Props> = (props: Props) => {
    const {tiles, coords} = props;
    const [tileDragEmitter, cellDropEmitter] = [
        useTileDragEmitter(),
        useCellDropEmitter(),
    ];
    const isValidMove = (validMoves: HexCoordinates[]) =>
        validMoves.some((dest) => deepEqual(coords, dest));
    const [classes, setClasses] = useState(['hex', 'cell']);
    const [currentTile, setCurrentTile] = useState<TileType | null>(null);

    function handleDragLeave(ev: { stopPropagation: () => void }) {
        ev.stopPropagation();
        if (currentTile)    setClasses(classes.filter(e => e !== 'active'))
    }

    function handleDragEnter(ev: { stopPropagation: () => void }) {
        ev.stopPropagation();
        if (currentTile) setClasses([...classes, 'active']);
    }

    function handleClickEvent(ev: { stopPropagation: () => void }) {
        if (currentTile) {
            setClasses([...classes, 'active']);
            tileDragEmitter.emit({type: 'end', tile: currentTile})
        }
    }

    const handleTileEvent = (e: TileDragEvent) => {
        const valid = isValidMove(e.tile.moves);
        if (e.type === 'start' && valid) {
            setClasses([...classes, 'can-drop']);
            setCurrentTile(e.tile)
        } else if (e.type === "start") {
            setCurrentTile(e.tile)
        } else if (e.type === 'end') {
            if (valid && currentTile && classes.includes('active'))
                cellDropEmitter.emit({
                    type: 'drop',
                    move: {coords, tileId: currentTile.id},
                });
            setCurrentTile(null);
            setClasses(['hex', 'cell']);
        }
    };

    useEffect(() => {
        tileDragEmitter.add(handleTileEvent);
        return () => tileDragEmitter.remove(handleTileEvent);
    });

    const attributes = {
        className: classes.join(' '),
        ondragover: handleDragOver,
        ondragleave: handleDragLeave,
        ondragenter: handleDragEnter,
        onmouseenter: handleDragEnter,
        onmouseleave: handleDragLeave,
        onclick: handleClickEvent,
    };

    return (
        <div {...attributes}>
            {tiles.length > 0 && <Tile {...tiles[0]} />}
        </div>
    );
};

CellFC.displayName = 'Cell';
export default memo(
    CellFC,
    (p, n) => deepEqual(p.coords, n.coords) && !p.tiles.length && !n.tiles.length
);
