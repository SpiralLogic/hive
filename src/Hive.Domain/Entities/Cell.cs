using System.Collections.Generic;
using System.Linq;

namespace Hive.Domain.Entities;

public sealed record Cell(Coords Coords)
{
    public Stack<Tile> Tiles { get; init; } = new();
    private readonly int _hashCode = Coords.GetHashCode();

    public bool Equals(Cell? other)
    {
        if (ReferenceEquals(other, null)) return false;
        return ReferenceEquals(this, other) || Coords.Equals(other.Coords);
    }

    public Cell AddTile(Tile tile)
    {
        Tiles.Push(tile);
        return this;
    }

    public bool IsEmpty()
    {
        return !Tiles.Any();
    }

    public Tile TopTile()
    {
        return Tiles.Peek();
    }

    public Tile RemoveTopTile()
    {
        return Tiles.Pop();
    }

    public override int GetHashCode() =>
        _hashCode;
}
