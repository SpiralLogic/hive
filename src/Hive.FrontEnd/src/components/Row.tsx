import '../css/row.css';

import { FunctionComponent } from 'preact';

type Props = { class?: string; zIndex?: number; hidden?: boolean };

const Row: FunctionComponent<Props> = (props) => {
  const { hidden, zIndex } = props;
  return (
    <div class={props.class} role={hidden ? 'none' : 'row'} style={zIndex ? { zIndex } : undefined}>
      {props.children}
    </div>
  );
};

Row.displayName = 'Row';
export default Row;
