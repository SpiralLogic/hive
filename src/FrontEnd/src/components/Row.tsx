import { deepEqual } from 'fast-equals';
import { FunctionComponent, h } from 'preact';
import { memo } from 'preact/compat';
import { HexCoordinates } from '../domain';
import Cell from './Cell';

type CellProps = typeof Cell.arguments['props'];
const cellKey = ({ q, r }: HexCoordinates) => `${q}-${r}`;
const createCell = (cell: CellProps) => <Cell key={cellKey(cell.coords)} {...cell} />;
const createPlaceholder = (key: number) => <div key={key} className="hidden" />;

type Props = {
    id: number;
    row: Array<CellProps>;
};

const Row: FunctionComponent<Props> = (props: Props) => {
    const { row } = props;
    const cells = row.map((cell, i) => (cell && createCell(cell)) || createPlaceholder(i));

    return <div className="hex-row">{cells}</div>;
};

Row.displayName = 'Row';
export default memo(Row, deepEqual);
