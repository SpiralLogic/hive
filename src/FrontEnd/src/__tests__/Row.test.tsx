import { h } from 'preact';
import { renderElement } from './helpers';
import Cell from '../components/Cell';
import Row from '../components/Row';

require('@testing-library/jest-dom');
jest.mock('fast-equals');

let row: HTMLElement;

describe('Row Tests', () => {
  beforeEach(() => {
    const cells = [
      { coords: { q: 0, r: 1 }, tiles: [], isHidden: true },
      { coords: { q: 0, r: 1 }, tiles: [] },
      { coords: { q: 1, r: 1 }, tiles: [] },
      { coords: { q: 0, r: 1 }, tiles: [], isHidden: true },
    ];
    row = renderElement(
      <Row
        children={cells.map((c) => (
          <Cell {...c} />
        ))}
      />
    );
  });

  test('renders multiple cells', () => {
    expect(row.children).toHaveLength(4);
  });

  describe('Row snapshot tests', () => {
    test('Row snapshot', () => {
      expect(row).toMatchSnapshot();
    });
  });
});
