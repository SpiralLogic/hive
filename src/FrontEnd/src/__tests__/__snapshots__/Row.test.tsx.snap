// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row Row tests Row snapshot 1`] = `
<div
  class="hex-row"
>
  <div
    class="hex cell"
  />
  <div
    class="hex cell"
  />
  <div
    class="hidden"
  />
</div>
`;
